File: src/host/lua/rvm.lua
local input = ");'u?>vD?>vRD?>vRA?>vRA?>vR:?>vR=!(:lkm!':lkv6y" -- RVM code that prints HELLO!

-- All variables are named such that the 'local' declarations can be
-- removed without causing clashes.  This rreduces the output Lua size
-- after minification, but makes LuaJIT less effective.

-- Functionality exposed to test wrapper.
local function rib(x,y,z) return {x,y,z} end
local function rib(x,y,z) --debug--
   assert(x and y and z) --debug--
   return { class = 'rib', [1]=x, [2]=y, [3]=z } --debug--
end --debug--

local FALSE=rib(0,0,5)
local TRUE=rib(0,0,5)
local NIL=rib(0,0,5)
local function is_rib(x) return type(x) == "table" end
-- local function is_rib(x) return type(x) == 'table' and x.class == 'rib' end --debug--

-- Called from test wrapper.  In normal use, this code falls into the
-- interpreter, and expects input to be defined.
local function test(input, dbg)  --debug--


local trace_instruction = function(...) end --debug--
if dbg then trace_instruction = dbg.trace_instruction end --debug--

local function putchar(x)
   local a=io.stdout
   a:write(string.char(x))
   a:flush()
   return x
end

local stack=0
local function push(x)
   stack = rib(x,stack,0)
end
local function getchar()
   local c=io.stdin:read(1)
   if not c then c=-1 else c=c:byte(1) end
   push(c)
end

local pos=0
local function get_byte()
   pos=pos+1
   local a=input:byte(pos) or 0
   return a
end

local function to_bool(x)
   if x then
      return TRUE
   else
      return FALSE
   end
end

local function pop()
   local x=stack[1]
   stack=stack[2]
   assert(x) --debug--
   return x
end

-- Argument evaluation order is undefined in Lua, vs. Python left-to-right.
-- Since we need to be explicit anyway, we use normal argument order for f,
-- as opposed to Python RVM which has them reversed.
local function prim1(f)
   return function()
      local a=pop()
      push(f(a))
   end
end
local function prim2(f)
   return function()
      local b=pop()
      local a=pop()
      push(f(a,b))
   end
end
local function prim3(f)
   return function()
      local c=pop()
      local b=pop()
      local a=pop()
      push(f(a,b,c))
   end
end

local function arg2()
      local a = pop()
      pop()
      push(a)
end
local function close()
      push(rib(pop()[1],stack,1))
end
local function f0s(x,y)
      x[1]=y
      return y
end
local function f1s(x,y)
      x[2]=y
      return y
end
local function f2s(x,y)
      x[3]=y
      return y
end

local function quotient(x, y)
   local b = math.floor
   local a = x/y
   if a<0 then
      return -b(-a)
   else
      return b(a)
   end
end

local primitives = {
   prim3(rib), -- 0
   prim1(function(x) return x end), -- 1
   pop, -- 2
   arg2, -- 3
   close, -- 4
   prim1(function(x) return to_bool(is_rib(x)) end), -- 5
   prim1(function(x) return x[1] end), -- 6
   prim1(function(x) return x[2] end), -- 7
   prim1(function(x) return x[3] end), -- 8
   prim2(f0s), -- 9
   prim2(f1s), -- 10
   prim2(f2s), -- 11
   prim2(function(x,y) return to_bool(x == y) end), -- 12
   prim2(function(x,y) return to_bool(x<y) end), -- 13
   prim2(function(x,y) return x + y end), -- 14
   prim2(function(x,y) return x - y end), -- 15
   prim2(function(x,y) return x * y end), -- 16
   prim2(quotient), -- 17
   getchar, -- 18
   prim1(putchar), -- 19
   prim1(os.exit) -- 20
}

local function get_code()
   local a = get_byte() - 35
   if a<0 then
      return 57
   else
      return a
   end
end

local function get_int(x)
   local a = get_code()
   x = x * 46
   if a<46 then
      return x+a
   else
      return get_int(x + a -46)
   end
end

local function list_tail(x, y)
   if y==0 then
      return x
   else
      return list_tail(x[2],y-1)
   end
end

local symtbl = NIL


local function symbol_ref(x)
   return list_tail(symtbl,x)[1]
end


local function set_global(x)
   symtbl[1][1]=x
   symtbl=symtbl[2]
end

-- build symtbl

-- Note that this variable is used to pass data between subsequent
-- parts of the code, similar to the Python code.
local n = get_int(0)

while n>0 do
   n = n - 1
   symtbl = rib(rib(0,rib(NIL,0,3),2),symtbl,0)
end
local accum = NIL
n=0
while true do
   local c = get_byte()
   if c == 44 then
      symtbl = rib(rib(0,rib(accum,n,3),2),symtbl,0)
      accum = NIL
      n = 0
   else
      if c == 59 then break end
      accum = rib(c, accum, 0)
      n = n + 1
   end
end
symtbl=rib(rib(0,rib(accum,n,3),2),symtbl,0)

-- n is unused

-- decode
while true do
   local x=get_code()
   n=x
   local d=0
   local op=0
   while true do
      local a={20,30,0,10,11,4}
      d=a[op+1]
      if n<=2+d then break end
      n=n-(d+3) ; op=op+1
   end
   if x>90 then
      n=pop()
   else
      if op==0 then
         stack=rib(0,stack,0);
         op = op + 1
      end
      if n==d then
         n = get_int(0)
      elseif n>= d then
         n = symbol_ref(get_int(n-d-1))
      elseif op<3 then
         n = symbol_ref(n)
      end
      if 4<op then
         n=rib(rib(n,0,pop()),0,1)
         if not is_rib(stack) then
            break
         end
         op=4
      end
   end
   stack[1]=rib(op-1,n,stack[1])
end

-- n contains main-proc, e.g. result of (decode) in hosts/scm/rvm.scm

set_global(rib(0,symtbl,1)) -- procedure type, primitive 0
set_global(FALSE)
set_global(TRUE)
set_global(NIL)

local function get_opnd(o)
   if is_rib(o) then return o
   else return list_tail(stack,o) end
end

local function get_cont()
   local a = stack
   while not is_rib(a[3]) do a = a[2] end
   return a
end

local pc = n[1][3]
stack=rib(0,0,rib(5,0,0)) -- primordial continuation (executes halt instr.)

-- run
while true do
   local o=pc[2]
   local i=pc[1]
   if i<1 then -- jump/call
      trace_instruction("jump/call",o,stack) --debug--
      o=get_opnd(o)[1]
      local c=o[1]
      if is_rib(c) then
         local c2=rib(0,o,0)
         local s2=c2
         nargs=c[1]
         while nargs > 0 do
            s2=rib(pop(),s2,0)
            nargs=nargs-1
         end
         if is_rib(pc[3]) then -- call
            c2[1]=stack
            c2[3]=pc[3]
         else -- jump
            k=get_cont()
            c2[1]=k[1]
            c2[3]=k[3]
         end
         stack=s2

      else
         primitives[c+1]()
         if is_rib(pc[3]) then -- call
            c=pc
         else --  jump
            c=get_cont()
            stack[2]=c[1]
         end
      end
      pc=c[3]

   elseif i<2 then -- set
      trace_instruction("set",o,stack) --debug--
      x=pop()
      get_opnd(o)[1]=x
      pc=pc[3]

   elseif i<3 then -- get
      trace_instruction("get",o,stack) --debug--
      push(get_opnd(o)[1])
      pc=pc[3]

   elseif i<4 then -- const
      trace_instruction("const",o,stack) --debug--
      push(o)
      pc=pc[3]

   elseif i<5 then -- if
      trace_instruction("if",o,stack) --debug--
      if pop() == FALSE then
         pc=pc[3]
      else
         pc=pc[2]
      end

   else -- halt
      break
   end
end

end --debug--

-- https://stackoverflow.com/questions/49375638/how-to-determine-whether-my-code-is-running-in-a-lua-module
-- Checks if there is anything in the 1st variable at the 4th level,
-- which would be the caller of the current module (if required) or
-- nothing in the case of the main script.
if pcall(debug.getlocal, 4, 1) then --debug--
   -- The debug code is part of lure luarocks package as
   -- 'lure.rvm_debug'.  It will load this code as a module, so defer
   -- exeuction here.
   return { test = test, is_rib = is_rib, rib = rib, FALSE = FALSE, TRUE = TRUE, NIL = NIL } --debug--
else test(input) end --debug--

File: src/host/scala/rvm.scala
import scala.concurrent.ExecutionContext
import scala.io.StdIn.readLine
import scala.collection.mutable.ArrayBuffer
import scala.sys.exit

enum ObjectType(val code: Int) {
    case Pair extends ObjectType(0)
    case Procedure extends ObjectType(1)
    case Symbol extends ObjectType(2)
    case String extends ObjectType(3)
    case Vector extends ObjectType(4)
    case SpecialValue extends ObjectType(5)
}

enum Opcode(val code: Int) {
    case JumpCall extends Opcode(0)
    case Set extends Opcode(1)
    case Get extends Opcode(2)
    case Const extends Opcode(3)
    case If extends Opcode(4)
}

/**
 * Represents one of the two value types that can be used within ribs.
 */
type RibField = Int | Rib

extension (f: RibField) {
    def asRib = f.asInstanceOf[Rib]
    def asInt = f.asInstanceOf[Int]
    def isRib = f.isInstanceOf[Rib]
    def isInt = f.isInstanceOf[Int]    
}

/**
 * Constant-size unit of memory used by the RVM.
 * Each field is either an integer or a reference to a rib.
 */
case class Rib(var x: RibField, var y: RibField, var z: RibField) {
    private var id =
        val idTemp = Rib.ribCounter
        Rib.ribCounter += 1
        idTemp
    
    override def toString: String =
        def toIntOrHashCode(v: RibField) = v match
            case v: Int => v
            case v: Rib => "["+v.hashCode+"]"
            "Rib<" + hashCode + ">("
                + toIntOrHashCode(x) + ", "
                + toIntOrHashCode(y) + ", "
                + toIntOrHashCode(z) + ")"

    override def hashCode: Int = id    
}

object Rib {
    private var ribCounter = 0

    def createSpecialValue() =
        Rib(0, 0, ObjectType.SpecialValue.code)
}

object Rvm {
    // By default, prints "HELLO!"  
    private val program = ");'u?>vD?>vRD?>vRA?>vRA?>vR:?>vR=!(:lkm!':lkv6y"

    private val DEBUG = false
    private val NB_OF_PRIMITIVES = 20

    private val falseRib = Rib.createSpecialValue()
    private val trueRib  = Rib.createSpecialValue()
    private val nilRib   = Rib.createSpecialValue()

    private var pos = 0
    private var symtable = nilRib    
    private var topOfStack: RibField = 0
    private var pc = Rib(0,0,0)

    private var inputBuffer = ""

    def main(args: Array[String]) =
        decodeSymbolTable()
        decodeInstructionGraph()
        setupGlobals()
        topOfStack = Rib(0, 0, Rib(5,0,0)) // Primordial continuation
        run()

    private def getByte() =
        val byte = program.charAt(pos).toByte
        pos += 1
        byte

    private def getCode() =
        val x = getByte() - 35
        if x < 0 then 57 else x
    
    private def getInt(n: Int): Int =
        val x = getCode()
        val n2 = n * 46
        if x < 46 then n2 + x else getInt(n2 + x - 46)
    
    private def listTail(rf: RibField, i: Int): RibField =           
        if (i == 0) rf else listTail( rf.asRib.y, i - 1)      
    
    private def decodeSymbolTable(): Unit =
        // Symbols with empty names
        val nbEmptySyms = getInt(0)          
        for _ <- 0 to (nbEmptySyms - 1) do
            symtable = Rib(Rib(0, Rib(nilRib, 0, 3), 2), symtable, 0)
        
        // Building the symbol table
        var n = 0
        var c: Byte = 0
        var accum = nilRib
        while c != 59 do
            c = getByte()              
            if c == 44 then
                symtable = Rib(Rib(0, Rib(accum, n, 3), 2), symtable, 0)
                accum = nilRib
                n = 0
            else if c != 59 then
                accum = Rib(c, accum, 0)
                n += 1

        symtable = Rib(Rib(0, Rib(accum, n, 3), 2), symtable, 0)

    private def decodeInstructionGraph(): Unit =
        def symbolRef(n: Int): RibField =
            listTail(symtable, n).asRib.x

        var m: RibField = 0

        var continue = true        
        while continue do
            val x = getCode()
            val shortEncodings = Array(20, 30, 0, 10, 11, 4)
            var n = x
            var d = 0
            var op = -1
            m = 0

            while
                op += 1
                d = shortEncodings(op)
                d + 2 < n
            do
                n -= d + 3
              
            if x > 90 then
                m = pop()
            else
                if op == 0 then
                    topOfStack = Rib(0, topOfStack, 0)
                    op+=1
                
                m = if n == d then getInt(0)
                    else if n > d then symbolRef(getInt(n - d - 1))
                    else if op < 3 then symbolRef(n)
                    else n
                
                if op > 4 then
                    m = Rib(Rib(m, 0, pop()), 0, 1)                    
                    if topOfStack == 0 then                                          
                        continue = false
                    else
                        op = 4
            
            if continue then              
                topOfStack.asRib.x = Rib(op-1, m, topOfStack.asRib.x)

        pc = m.asRib.x.asRib.z.asRib
        
    private def setupGlobals(): Unit =
        def setGlobal(r: Rib): Unit =
            symtable.x.asRib.x = r
            symtable = symtable.y.asRib
        
        setGlobal(Rib(0, symtable, ObjectType.Procedure.code))
        setGlobal(falseRib)
        setGlobal(trueRib)
        setGlobal(nilRib)       
    
    /*
     * [Execution]
     */
    private def pop(): RibField =        
        val popped = topOfStack.asRib.x        
        topOfStack = topOfStack.asRib.y
        popped

    private def push(x: RibField): Unit =        
       topOfStack = Rib( x , topOfStack, 0)

    private def getOpnd(rf: RibField): RibField =        
        if(rf.isRib) rf.asRib
        else listTail(topOfStack.asRib, rf.asInt)

    private def getCont(): Rib = 
        var s = topOfStack.asRib        
        while{ (s.z.isInt && s.z.asInt == 0) && s.y.isRib } do
            s = s.asRib.y.asRib        
        s

    private def putChar(c: RibField): RibField =         
        print( c.asInt.toChar )
        c

    private def toBool(b: Boolean): Rib =       
        if(b) trueRib else falseRib

    private def prim1( f:(RibField) => RibField ): Unit = 
        push( f( pop() ) )
    
    private def prim2( f:(RibField, RibField) => RibField ): Unit = 
        push( f( pop(), pop() ) )

    private def prim3( f:(RibField, RibField, RibField) => RibField ): Unit =        
        push( f( pop(), pop(), pop() ) )

    private def primitives(no: Int): Unit = no match {
        case 0 => prim3( (x, y, z) => Rib(z, y, x) ) // Rib
        case 1 => prim1( (rf: RibField) => rf )      // ID
        case 2 => pop() // Pop
        case 3 => {     // Skip
            val y = pop()
            pop()
            push(y)
        }
        case 4  => push( Rib( pop().asRib.x, topOfStack, 1) ) // Close
        case 5  => prim1( (x: RibField) => toBool(x.isRib) )  // isRib ?
        case 6  => prim1( (x: RibField) => x.asRib.x ) // Field0
        case 7  => prim1( (x: RibField) => x.asRib.y ) // Field1
        case 8  => prim1( (x: RibField) => x.asRib.z ) // Field2
        case 9  => prim2( (y: RibField, x: RibField) => { // Set Field0(x)
            x.asRib.x = y
            y
        })
        case 10 => prim2( (y: RibField, x: RibField) => { // Set Field1(y)
            x.asRib.y = y
            y
        }) 
        case 11 => prim2( (y: RibField, x: RibField) => { // Set Field2(z)
            x.asRib.z = y
            y
        }) 
        case 12 => prim2( (y: RibField, x: RibField) => toBool( if(x.isRib && y.isRib) x.asRib eq y.asRib else x == y ) ) // Equal    
        case 13 => prim2( (y: RibField, x: RibField) => toBool( x.asInt < y.asInt ) ) // Less than
        case 14 => prim2( (y: RibField, x: RibField) => x.asInt + y.asInt ) // Add
        case 15 => prim2( (y: RibField, x: RibField) => x.asInt - y.asInt ) // Sub
        case 16 => prim2( (y: RibField, x: RibField) => x.asInt * y.asInt ) // Mult
        case 17 => prim2( (y: RibField, x: RibField) => if(y.asInt != 0) x.asInt/y.asInt else 0 ) // Div         
        case 18 => { // GetChar
            if inputBuffer == "" then
                inputBuffer = readLine()
                if inputBuffer != null then
                    inputBuffer += "\n"
                else
                    inputBuffer = ""
            
            if inputBuffer == "" then
                push(-1)
            else
                val retChar = inputBuffer.charAt(0)
                inputBuffer = inputBuffer.substring(1)
                push(retChar)
        }
        case 19 => prim1( putChar )// PutChar
        case 20 => { exit() }
    }
 
    private def run(): Unit = {  
        var o = pc.y      
        val instr = pc.x.asInt   
        var nextPC = true

        instr match{ 
            case 5 => return // Halt
            case 0 => {      // Jump/Call    
                o = getOpnd(o).asRib.x                   
                var c = o.asRib.x   
               
                if( c.isRib ){                    
                    var c2 = Rib(0,o,0)                   
                    var s2 = c2    
                    var nargs = c.asRib.x.asInt                   
                    while(nargs > 0)                        
                        s2 = Rib( pop(), s2, 0)                       
                        nargs -= 1 
                                                         
                    if(pc.z.isRib){    
                        c2.x = topOfStack
                        c2.z = pc.z  
                    }
                    else{                        
                        var k = getCont()                       
                        c2.x = k.x
                        c2.z = k.z  
                    }    
                                
                    topOfStack = s2
                }
                else{                                             
                    if(c.asInt <= NB_OF_PRIMITIVES) primitives(c.asInt) else return               
                                                             
                    if(pc.z.isRib) c = pc  
                    else{
                        c = getCont()
                        topOfStack.asRib.y = c.asRib.x                                                 
                    }                                                    
                }  
                
                pc = c.asRib.z.asRib  
                nextPC = false                                   
            }
            case 1 => getOpnd(o).asRib.x = pop() // Set    
            case 2 => // Get               
                val t = getOpnd(o) 
                val k = t.asRib.x               
                push(k)                  
            case 3 =>  push(o) // Const   
            case 4 =>          // If                            
                if
                    val popped = pop()                    
                    // Everything is true except for `falseRib`
                    !( popped.isRib && (popped.asRib eq falseRib ) )                   
                then  
                    pc = pc.y.asRib  
                    nextPC = false             
        }  

        if(nextPC) pc = pc.z.asRib  
        run() // Tail recursion  
    }   
}